### PR TITLE
16b Target Model Converters

### PR DESCRIPTION
**Pull request by tpolecat**
This adds code to the `To2016B` program converter to generate new-model targets from old ones and add a zero-length scheduling block to nonsidereal observations, corresponding with the single-element ephemeris generated from the existing fixed coordinates and `validAt` time. Right now the deserialized values contain *both* old and new targets; the old target will be removed once it is unused.

### CODE REVIEW

**Comment by cquiroz** [comment]
The error on travis is the same we observed on my previous PR #848. I marked the broken tests as pending

File: bundle/edu.gemini.spModel.io/src/main/scala/edu/gemini/spModel/io/impl/migration/to2016B/To2016B.scala [hunk 1]
```diff
@@ -68,4 +81,116 @@ object To2016B extends Migration {
...
+      // Drop it into the paramset.
+      // TODO: remove everything else!
+      t.addParamSet(TargetParamSetCodecs.TargetParamSetCodec.encode("target", newTarget))
+
+    }
+
+  // Properties common to all target types
+  def common(ps: ParamSet): State[Target, Unit] =
+    for {
+      _ <- Target.name                ?:= ps.value("name")
```
Comment (cquiroz) [inline bundle/edu.gemini.spModel.io/src/main/scala/edu/gemini/spModel/io/impl/migration/to2016B/To2016B.scala]: Can you explain what the `?:=` and `:=` operators do?

File: bundle/edu.gemini.spModel.io/src/test/scala/edu/gemini/spModel/io/impl/migration/MigrationTest.scala [hunk 2]
```diff
@@ -21,11 +20,46 @@ trait MigrationTest {
...
+      val parser = new PioSpXmlParser(db.getFactory)
+      val files  = Option(dir.listFiles).getOrElse(Array.empty[File]).filter(_.getName.endsWith(".xml")).sortBy(_.getName)
+      val total  = files.length
+      files.zipWithIndex.foreach { case (f, n) =>
+        print(s"[$n/$total] ${f.getName} ... ")
+        try {
+          parser.parseDocument(new FileReader(f))
+          println(" [ok]")
+        } catch {
+          case e: Exception => println(s" ${Console.RED}[err] ${e.getMessage}${Console.RESET}")
```
Comment (cquiroz) [inline bundle/edu.gemini.spModel.io/src/test/scala/edu/gemini/spModel/io/impl/migration/MigrationTest.scala]: Is the `Console.RED` used to change color in the output? ANSI codes?

Comment (tpolecat) [inline bundle/edu.gemini.spModel.io/src/test/scala/edu/gemini/spModel/io/impl/migration/MigrationTest.scala]: Yeah that's all it does. It's just ANSI escapes.

File: bundle/edu.gemini.spModel.io/src/main/scala/edu/gemini/spModel/io/impl/migration/to2016B/To2016B.scala (hunk 1 above)
Comment (tpolecat) [inline bundle/edu.gemini.spModel.io/src/main/scala/edu/gemini/spModel/io/impl/migration/to2016B/To2016B.scala]: Yeah `:=` turns a lens into a state action:  `Lens[A,B] := B` gives you a `State[A, B]`. So this lets you use imperative-looking code to write a pure functional series of lens transformations. So I compose all the "assignments" into a single `State` program and then use `exec` to apply it to an `A`.

`?:=` is defined up at the top of the file. It does the same thing but for `Lens[A,B] := Option[B]` that is a no-op if the value is empty.

File: bundle/edu.gemini.spModel.io/src/main/scala/edu/gemini/spModel/io/impl/migration/to2016B/To2016B.scala [hunk 3]
```diff
@@ -68,4 +81,116 @@ object To2016B extends Migration {
...
+        for {
+          system <- t.value("system")
+          ra     <- t.value("c1").flatMap(s => (Angle.parseHMS(s) orElse Angle.parseDegrees(s)).toOption).map(RightAscension.fromAngle)
+          dec    <- t.value("c2").flatMap(s => (Angle.parseDMS(s) orElse Angle.parseDegrees(s)).toOption).flatMap(Declination.fromAngle)
+        } yield (system, Coordinates(ra, dec))
+
+      // Construct a new target
+      val newTarget: Target =
+        data.map {
+          case ("J2000", Coordinates.zero) => TooTarget.empty
```
Comment (swalker2m) [inline bundle/edu.gemini.spModel.io/src/main/scala/edu/gemini/spModel/io/impl/migration/to2016B/To2016B.scala]: By default a new target in a non-ToO observation is also Sidereal (0,0).  To be sure, you would need to also check the ToO type of the program.

Comment (tpolecat) [inline bundle/edu.gemini.spModel.io/src/main/scala/edu/gemini/spModel/io/impl/migration/to2016B/To2016B.scala]: Ah that's right. Thanks.

**Comment by cquiroz** [comment]
If you rebase on develop it should pass travis checks

**Comment by tpolecat** [comment]
@cquiroz yep, will do.

**Comment by tpolecat** [comment]
@swalker2m ok I fixed that TOO issue and added a test for it.

**Comment by swalker2m** [comment]
Looks good.  :+1:

**Pull request merged**